### ISSUE TITLE
[Windows/arm32] Test failure: 10w5d_cs_do.cmd 

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/coreclr#18988]
With JitStress=1

https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm/job/jitstress/job/arm_cross_checked_windows_nt_jitstress1_tst/33/

```
TEST_CMD_LINE: JIT\Methodical\fp\exgen\10w5d_cs_do> C:\Users\robox\j\workspace\arm_cross_che---9b14f1d9\bin\tests\Windows_NT.arm.Checked\JIT\Methodical\fp\exgen\10w5d_cs_do\10w5d_cs_do.cmd 

BEGIN EXECUTION
 "C:\Users\robox\j\workspace\arm_cross_che---9b14f1d9\bin\tests\Windows_NT.arm.Checked\Tests\Core_Root\corerun.exe" 10w5d_cs_do.exe 
Func_0_6_5_6_3: != true
Func_0_6_5_4_6: <= true
Func_0_6_5_4_5: > true
Func_0_6_5_4_3: == false
Func_0_6_4_6_2: != true
Func_0_6_4_5_4: < true
Func_0_6_4_5_1: < true
Func_0_6_4_4: <= true
Func_0_6_3_6_5: == false
Func_0_6_3_4_4: <= true
Func_0_6_3_4_3: < true
Func_0_6_3_3: < true
Func_0_6_3_2_3: >= true
Func_0_6_3_1_5: < true
Func_0_6_3_1: <= false
Func_0_6_3: != true
Func_0_6_2_5: != true
Func_0_6_2_1: > false
Func_0_6_1_6_4: < false
Func_0_6_1_6_3: > false
Func_0_6_1_5_4: > true
Func_0_6_1_3: == false
Func_0_5_6_6_6: <= true
Func_0_5_6_6_1: > true
Func_0_5_6_4_6: <= false
Func_0_5_6_4_2: >= false
Func_0_5_6_3_6: < false
Func_0_5_6_3_1: == false
Func_0_5_6_3: <= true
Func_0_5_6_2_1: >= false
Func_0_5_6_1_3: > true
Func_0_5_6_1: <= false
Func_0_5_5_6: != true
Func_0_5_5_5_6: != true
Func_0_5_5_5_4: == false
Func_0_5_5_5_2: == false
Func_0_5_5_4_5: < true
Func_0_5_5_4: > true
Func_0_5_5_3_4: != true
Func_0_5_5_3_2: <= true
Func_0_5_5_3_1: > true
Func_0_5_5_1: < false
Func_0_5_4_5: >= false
Func_0_5_4_1_3: <= false
Func_0_5_4_1_2: == false
Func_0_5_4: != true
Func_0_5_3_6: != true
Func_0_5_3_5_3: < true
Func_0_5_3_4_2: == false
Func_0_5_3_4_1: >= true
Func_0_5_3_4: <= true
Func_0_5_3_1_2: < true
Func_0_5_3: >= false
Func_0_5_2_6_6: != true
Func_0_5_2_6_3: != true
Func_0_5_2_5_6: != true
Func_0_5_2_5_4: == false
Func_0_5_2_5_3: <= true
Func_0_5_2_5_2: != true
Func_0_5_2_4_5: > false
Func_0_5_2_4_2: <= true
Func_0_5_2_2_5: > true
Func_0_5_2_1_6: > false
Func_0_5_2_1_1: != true
Func_0_5_1: != true
Func_0_4_6_6_5: <= true
Func_0_4_6_6: == false
Func_0_4_6_5_6: <= true
Func_0_4_6_5_5: != true
Func_0_4_6_5_2: <= true
Func_0_4_6_5: == false
Func_0_4_6_3_6: >= true
Func_0_4_6_3_3: == false
Func_0_4_6_3_2: <= false
Func_0_4_6_2: == false
Func_0_4_5_6: >= false
Func_0_4_5_5: < true
Func_0_4_5_3_5: == false
Func_0_4_5_2_6: > false
Func_0_4_5_2_4: >= true
Func_0_4_5_2_1: <= false
Func_0_4_5_2: >= true
Func_0_4_5_1_5: == false
Func_0_4_5_1_2: > true
Func_0_4_4_5: != true
Func_0_4_4_2_6: < true
Func_0_4_4_2_5: != false
Func_0_4_4_2_3: <= true
Func_0_4_4_2: == false
Func_0_4_4_1_4: >= false
Func_0_4_3_6_6: <= true
Func_0_4_3_6_5: <= true
Func_0_4_3_6_2: >= true
Func_0_4_3_6_1: == true
Func_0_4_3_4: >= true
Func_0_4_3_2_2: == false
Func_0_4_3_1_3: > true
Func_0_4_3_1_1: != true
Func_0_4_2_5_4: <= true
Func_0_4_2_5_1: != true
Func_0_4_1_4: >= true
Func_0_4_1_3_1: >= true
Func_0_4_1_2_3: < false
Func_0_4_1_1_1: >= true
Func_0_3_6_6_5: >= true
Func_0_3_6_6_1: < true
Func_0_3_6_5_5: > false
Func_0_3_6_4_1: < false
Func_0_3_6_3_4: != true
Func_0_3_5_6_6: == true
Func_0_3_5_6_5: <= true
Func_0_3_5_6_2: != true
Func_0_3_5_4: == false
Func_0_3_5_3: > false
Func_0_3_5_2_6: != true
Func_0_3_5_2_5: <= true
Func_0_3_5_2_4: > false
Func_0_3_5_2_2: == false
Func_0_3_5_1_1: != true
Func_0_3_4_1_1: < true
Func_0_3_4_1: < true
Func_0_3_3_6_4: <= false
Func_0_3_3_6_1: <= true
Func_0_3_3_5_6: != true
Func_0_3_3_5_1: != true
Func_0_3_3_4_6: <= true
Func_0_3_3_4_5: <= true
Func_0_3_3_4_2: < true
Func_0_3_2_6_6: <= true
Func_0_3_2_6_3: >= false
Func_0_3_2_6_1: >= true
Func_0_3_2_5_6: < true
Func_0_3_2_5_3: >= true
Func_0_3_2_5_1: > true
Func_0_3_2_4_3: > true
Func_0_3_2_3_3: > true
Func_0_3_2_3_2: <= false
Func_0_3_2_1_5: != true
Func_0_3_1_6_6: > true
Func_0_3_1_6_4: >= false
Func_0_3_1_4_4: <= true
Func_0_3_1_4_3: == true
Func_0_3_1_4_2: == false
Func_0_3_1_4_1: <= true
Func_0_3_1_3_5: == false
Func_0_3_1_3_4: < true
Func_0_3_1_2_3: < true
Func_0_2_6_6_4: != false
Func_0_2_6_6_3: < true
Func_0_2_6_6_2: != true
Func_0_2_6_5_2: > true
Func_0_2_6_4_5: != true
Func_0_2_6_3_1: >= true
Func_0_2_6_1_3: <= true
Func_0_2_6_1_2: == false
Func_0_2_5_6_6: == false
Func_0_2_5_6_1: >= false
Func_0_2_5_5_2: <= false
Func_0_2_5_5_1: != true
Func_0_2_5_4_3: <= true
Func_0_2_5_4_1: > true
Func_0_2_3_6_5: != true
Func_0_2_3_6_4: <= true
Func_0_2_3_6_1: >= false
Func_0_2_3_4_5: < true
Func_0_2_3_4_3: < false
Func_0_2_3_3_4: == false
Func_0_2_3_3_1: < true
Func_0_2_3_2_4: < true
Func_0_2_3_1_6: <= true
Func_0_2_3_1_2: <= false
Func_0_2_3_1_1: >= true
Func_0_2_3_1: == true
Func_0_2_2_6: != true
Func_0_2_2_3_5: > false
Func_0_2_2_3_3: != true
Func_0_2_2_3_1: < true
Func_0_2_2_2: != true
Func_0_2_2_1_5: > true
Func_0_2_2_1_3: < true
Func_0_2_2_1_2: > true
Func_0_2_2_1: < true
Func_0_2_1_2_2: >= true
Func_0_1_6_6_5: == false
Func_0_1_6_2_4: < true
Func_0_1_6_2_2: <= true
Func_0_1_6_1_6: != true
Func_0_1_6_1_3: > false
Func_0_1_6_1_2: != true
Func_0_1_6: >= true
Func_0_1_5_5_1: != true
Func_0_1_5_4_6: == false
Func_0_1_5_4_5: == false
Func_0_1_5_4_4: < true
Func_0_1_5_4_1: == false
Func_0_1_5_3_3: == false
Func_0_1_4_2_3: > true
Func_0_1_1_6_5: > true
Func_0_1_1_6_3: >= false
Func_0_1_1_6_2: != true
Func_0_1_1_6_1: <= false
Func_0_1_1_5_2: != true
Func_0_1_1_2_6: < true
Func_0_1_1_2_4: > true
Func_0_1_1_2_3: == false
Func_0_1_1_2_1: < false
Func_0_1_1_1_5: >= true
Expected: 100
Actual: -532462766
END EXECUTION - FAILED
FAILED


TEXT_EXPECTED_RETURN_CODE = 0
TEXT_ACTUAL_RETURN_CODE = 1 (0x1)
```

**Comment by BruceForstall** [comment]
Looks like a regression in just this build.

**Comment by BruceForstall** [comment]
@dotnet/arm32-contrib @dotnet/jit-contrib

**Comment by sandreenko** [comment]
The regression appeared after #18948

**Comment by sandreenko** [comment]
Ok, the root case is the wrong implementation of `Decimal::Low64`:
https://github.com/dotnet/coreclr/blob/25452220f8486f9ecb56b7ffa84a37d80136b750/src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs#L24-L28

`Unsafe.As<int, ulong>(ref int a)` this cast produces wrong result, for example:
`int a = 5; (in little endian it is 50 00 00 00)`, then we have `int b = 0x10000 ( 00001)` sequentially in the memory , 
then `Unsafe.As<int, ulong>(ref int a)` will cast a pointer to int to a pointer to long and the cast result will have size of 8 bytes and obviously value of `a ## b == 1000000000005`,
so `Unsafe.As<int, ulong>(ref int a) == 1000000000005` and `Decimal.ToDouble(5M)` will return  trash value.

@pentp could you please explain where did this implementation come from? As I understand this work was done in CoreRT repo, but now it doesn't [show the history of this file](https://github.com/dotnet/corert/commits/master/src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs), it shows only the last change `Move Decimal to shared`. But as I see from[ this history ](https://github.com/dotnet/corert/commits/ba2feed8aa3db9fca21a00206b445d3080e0ac16/src/System.Private.CoreLib/src/System/Decimal.DecCalc.cs) that there was a correct implementation of `Low64`:
```
        private ulong Low64
        {
#if BIGENDIAN
            get { return ((ulong)umid << 32) | ulo; }
            set { umid = (uint)(value >> 32); ulo = (uint)value; }
#else
            get => ulomidLE;
            set => ulomidLE = value;
#endif
        }
```
but I can't find where and why it was changed, could you please point me?

cc @jkotas @dotnet/jit-contrib 

If my analysis is correct than we have a weak coverage for this class if this issue happens only with stress mode.

**Comment by sandreenko** [comment]
> The code takes advantage of the fact that the lo and mid fields are next to each other in the struct. 

Could we guarantee their layout according to .Net standard?

> We can write it differently if writing it this way is incompatible with the JIT optimizations.
As I understand we can make a struct promotional and it will volatile this layout fact.

**Comment by jkotas** [comment]
Does it still hit the problem if the implementation is changed to:

`private ulong Low64 => Unsafe.As<decimal, DecCalc>(ref Unsafe.AsRef(in this)).Low64;` 

and the DecCalc.Low64 property is changed to internal

?

**Comment by jkotas** [comment]
> Could we guarantee their layout according to .Net standard?

Structs in C# have sequential layout by default that guarantees that the fields are next to each other.

**Comment by AndyAyersMS** [comment]
The jit currently may miss detecting some cases where you walk off one field and onto another. See for instance #16472 (which looks pretty much like the issue here).

Having us pin the entire struct if a field's address is taken is likely too conservative as most address taking is harmless. So I think the plan is for us to shore up the address-exposure logic.

**Comment by sandreenko** [comment]
> Does it still hit the problem if the implementation is changed to:
> private ulong Low64 => Unsafe.As<decimal, DecCalc>(ref Unsafe.AsRef(in this)).Low64;
> and the DecCalc.Low64 property is changed to internal
> ?

No, it doesn't.

**Comment by sandreenko** [comment]
> The jit currently may miss detecting some cases where you walk off one field and onto another. See for instance #16472 (which looks pretty much like the issue here).
> Having us pin the entire struct if a field's address is taken is likely too conservative as most address taking is harmless. So I think the plan is for us to shore up the address-exposure logic.

Looks similar.

The jit dump shows:
```
lvaGrabTemp returning 98 (V98 tmp87) (a long lifetime temp) called for field V84.hi (fldOffset=0x4).
lvaGrabTemp returning 99 (V99 tmp88) (a long lifetime temp) called for field V84.lo (fldOffset=0x8).
lvaGrabTemp returning 100 (V100 tmp89) (a long lifetime temp) called for field V84.mid (fldOffset=0xc).
```
and then:
```
;  V98 tmp87        [V98,T56] (  2,  2   )     int  ->   r5         V84.hi(offs=0x04) P-INDEP
;  V99 tmp88        [V99,T57] (  2,  2   )     int  ->  [sp+0x34]   do-not-enreg[F] V84.lo(offs=0x08) P-INDEP
;* V100 tmp89       [V100    ] (  0,  0   )     int  ->  zero-ref    V84.mid(offs=0x0c) P-INDEP
```

as I understand `V100 tmp89` must be allocated as `[sp+0x38] ==  offset of (V99 tmp88) + 0x4` and it is jit error that it doesn't, is it correct?

**Comment by mikedn** [comment]
> as I understand V100 tmp89 must be allocated as [sp+0x38] == offset of (V99 tmp88) + 0x4 and it is jit error that it doesn't, is it correct?

The obvious problem is that the struct ended up being independent promoted. It should have been dependent promoted. Or better, only the `lo` and `mid` fields should be dependent and not enregistered. But I don't think the JIT currently supports this kind of mixed dependent/independent promotion.

Though this begs the question what this optimization is trying to achieve. If the struct is in already forced in memory for other reasons then this should be a good optimization. But for structs that are local variables and promotion candidates this optimization can hurt and even the best optimizer may not produce the desired results because there aren't many options:
* force the entire struct in memory - a fixed JIT would do just that, hurts all other "normal" struct accesses
* force only lo and mid in memory - that will still hurt normal accesses to those fields
* spilling lo and mid to memory right before the int64 read - nice but not that easy to implement and possible not that efficient depending on usage
* if lo and mid are already in registers just do the shift thing - "great", the JIT did what the developer was trying to avoid. But that probably good since the developer had no way of knowing that those fields are in registers.

I wonder if it wouldn't be slightly easier for the JIT to go the other way around - in the shift case figure out that both field are already in memory and change the shift to a int64 load.

**Comment by sandreenko** [comment]
Ok, thanks all for the help, the root case is that `fgMarkAddrTakenLocalsPreCB` works wrong for such case:
```
               [000410] *-CXG--------             |  |  \--*  IND       long  
               [000419] ----G--------             |  |     \--*  ADDR      byref 
               [000420] ----G--------             |  |        \--*  FIELD     int    lo
               [000421] L------------             |  |           \--*  ADDR      byref 
               [000422] -------------             |  |              \--*  LCL_VAR   struct V18 tmp13  
```
```
fgMarkAddrTakenLocalsPreCB for tree 410, value on stack: AXC_None, pushes: AXC_Ind
fgMarkAddrTakenLocalsPreCB for tree 419, value on stack: AXC_Ind, pushes: AXC_None
fgMarkAddrTakenLocalsPreCB for tree 420, value on stack: AXC_None, pushes: AXC_NONE

```
when `fgMarkAddrTakenLocalsPreCB` for `000420` should catch that we access it as wider type and set `lvaSetVarAddrExposed`.

I do not see an obvious solution for this right now, trying to repro it on x64.

**Comment by BruceForstall** [comment]
This is also failing for JitStress1 on arm32 Ubuntu:
https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm/job/jitstress/job/arm_cross_checked_ubuntu_jitstress1_flow/24/

@sandreenko Can you disable this test for all platforms (maybe just under JIT stress: mark it `<JitOptimizationSensitive>true</JitOptimizationSensitive>`) until a solution is implemented?

**Comment by sandreenko** [comment]
some examples of the wrong behavior:
1. when we set `addrExposed` when we do not need to do so:
1.1 `IND as int(COMMA(NOP, ADDR(LCL_VAR int))`, when `fgMarkAddrTakenLocalsPreCB` process `GT_COMMA` it reset `context` and then thinks that `LCL_VAR` addr is exposed'
1.2 'LCL_VAR 02= OBJ(LCL_VAR 01); ADDR(LCL_VAR 01)`, again LCL_VAR 01 will be set as addr exposed, because our current analysis doesn't see other stmts.

2. we do not set `addrExposed` when we need to do so:
2.1 `IND as long(ADDR(FIELD int))` <- current issue;

2.2           
```
               [002674] --CXG-------             / --*  CALL      ref IsByRefArray
               [002673] ---XG------- this in ecx | \--*  ADDR      byref
               [002671] ------------             |  |  /--*  CNS_INT   int    0
               [002672] ---XG-------             |  \--*  INDEX     struct
               [002670] ------------             |     \--*  LCL_VAR   ref    V06 arg6
```
the current analysis says `Taking the address of an array element never takes the address of a local.` that I think is not always correct.

**Comment by mikedn** [comment]
Are you trying to fix `fgMarkAddrTakenLocalsPreCB`? That thing needs a rewrite, not a fix.

> when we set addrExposed when we do not need to do so:

Beware of any change that makes it less conservative. Marking address taken locals is done under the assumption that morph will be able to eliminate the ADDR operation. If some lclvars are no longer marked and morph fails to eliminate the ADDR then you'll likely run into problems.

**Comment by sandreenko** [comment]
Fixed by @mikedn  #19621.
